### PR TITLE
pretty much rewrites lib

### PR DESCRIPTION
**Pull request by hernanat**
- use Tesla
  - make both http client & json lib configurable (see config)

- use singular module names

- still wip

### CODE REVIEW

**Comment by hernanat** [comment]
ok @houllette -- I ended up getting something working w/ Tesla if you want to pull this and play with it I'd be curious your thoughts. Still a WIP as there are some cleanup items I'd like to address. Also there are no tests yet

File: lib/open_ai/api_client.ex [hunk 1]
```diff
@@ -0,0 +1,234 @@
+defmodule OpenAI.ApiClient do
+  @client_impl Application.compile_env(:open_ai, [:impls, :api_client], OpenAI.ApiClient.Impl)
+
+  @typedoc """
+  API Client type. `http_config` contains configuration options for requests. By
+  default it will be a `t:Tesla.Client.t/0` since we use `Tesla`, but your own
```
Comment (hernanat) [inline lib/open_ai/api_client.ex:6]: the more I think about it the less I want to encourage people to write their own client impls. I think we will only have two: the live one, and the one we'll use for tests. All this behaviour / impl stuff is more than a user needs to know / care about

File: config/config.exs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import Config
+
+config :open_ai,
+       :application,
+       children: [{Finch, name: OpenAIFinch}]
```
Comment (hernanat) [inline config/config.exs:5]: one awkward thing that we'll need to remind users of if we want to keep finch as the default is that if they use something other than finch they need to make sure to get rid of this in their own config, i.e.:

```elixir
config :open_ai, :application, children: []
```

or st

File: config/config.exs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+config :open_ai, :tesla,
+  adapter: {Tesla.Adapter.Finch, [name: OpenAIFinch]},
+  retry: [
+    delay: 50,
+    max_retries: 3,
```
Comment (hernanat) [inline config/config.exs:11]: not sure if these are reasonable retry defaults. Don't really want to pay to have myself rate limited for an OSS project ... can revisit if/when there is community feedback lol

File: lib/open_ai/completion.ex [hunk 4]
```diff
@@ -1,38 +1,38 @@
...
-  The parameters are mapped 1:1 with those that OpenAI (except for streaming) offers and so
-  we do not explain them in detail here.
-
-  *Note*: The `:stream` param is not currently supported.
+  The parameters are mapped 1:1 with those that OpenAI offers and so we do not explain
+  them in detail here.
   """
   @type create_params ::
-          {:suffix, binary()}
+          {:suffix, String.t()}
```
Comment (hernanat) [inline lib/open_ai/completion.ex:23]: switching `binary()` -> `String.t()` as one of my coworkers kindly informed me this is a better practice when expecting UTF-8 encoded strings.

File: mix.exs [hunk 5]
```diff
@@ -31,10 +31,23 @@ defmodule OpenAI.MixProject do
...
     [
       {:ex_doc, "~> 0.29", only: :dev, runtime: false},
-      {:finch, "~> 0.14"},
-      {:hammox, "~> 0.7", only: :test},
-      {:jason, "~> 1.3"},
-      {:multipart, "~> 0.3"}
+      {:tesla, "~> 1.5"},
+
+      # http clients
+      {:ibrowse, "4.4.0", optional: true},
```
Comment (hernanat) [inline mix.exs:37]: I copied these versions from tesla's mix.exs file but I will revisit to see if they need to be adjusted

**Comment by houllette** [comment]
You weren't kidding! This is awesome - I'll be sure to dig in more here soon and provide some more actionable feedback. I haven't forgotten about this 😉

**Comment by hernanat** [comment]
this PR is deprecated -- I rewrote it again to make it more easily customizable

**Pull request closed without merging**